helper/testhelpers/certhelpers/cert_helpers.go

```diff
@@ -0,0 +1,258 @@
+package certhelpers
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"io/ioutil"
+	"math/big"
+	"net"
+	"os"
+	"strings"
+	"testing"
+	"time"
+)
+
+type CertBuilder struct {
+	tmpl       *x509.Certificate
+	parentTmpl *x509.Certificate
+
+	selfSign  bool
+	parentKey *rsa.PrivateKey
+
+	isCA bool
+}
+
+type CertOpt func(*CertBuilder) error
+
+func CommonName(cn string) CertOpt {
+	return func(builder *CertBuilder) error {
+		builder.tmpl.Subject.CommonName = cn
+		return nil
+	}
+}
+
+func Parent(parent Certificate) CertOpt {
+	return func(builder *CertBuilder) error {
+		builder.parentKey = parent.privKey.privKey
+		builder.parentTmpl = parent.Template
+		return nil
+	}
+}
+
+func IsCA(isCA bool) CertOpt {
+	return func(builder *CertBuilder) error {
+		builder.isCA = isCA
+		return nil
+	}
+}
+
+func SelfSign() CertOpt {
+	return func(builder *CertBuilder) error {
+		builder.selfSign = true
+		return nil
+	}
+}
+
+func Ip(ip ...string) CertOpt {
+	return func(builder *CertBuilder) error {
+		for _, addr := range ip {
+			if ipAddr := net.ParseIP(addr); ipAddr != nil {
+				builder.tmpl.IPAddresses = append(builder.tmpl.IPAddresses, ipAddr)
+			}
+		}
+		return nil
+	}
+}
+
+func Dns(dns ...string) CertOpt {
+	return func(builder *CertBuilder) error {
+		builder.tmpl.DNSNames = dns
+		return nil
+	}
+}
+
+func NewCert(t *testing.T, opts ...CertOpt) (cert Certificate) {
+	t.Helper()
+
+	builder := CertBuilder{
+		tmpl: &x509.Certificate{
+			SerialNumber: makeSerial(t),
+			Subject: pkix.Name{
+				CommonName: makeCommonName(),
+			},
+			NotBefore: time.Now().Add(-1 * time.Hour),
+			NotAfter:  time.Now().Add(1 * time.Hour),
+			IsCA:      false,
+			KeyUsage: x509.KeyUsageDigitalSignature |
+				x509.KeyUsageKeyEncipherment |
+				x509.KeyUsageKeyAgreement,
+			ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+			BasicConstraintsValid: true,
+		},
+	}
+
+	for _, opt := range opts {
+		err := opt(&builder)
+		if err != nil {
```

Comment:
If we're only evaluating a local variable, it's common to see these statements on one line like so:

```go
if err := opt(&builder); err != nil {
  // reference err here
}
```

Reply:
Interesting style, noted. Used elsewhere, unsure if I changed this one.